src/OpenLoco/src/GameCommands/CreateIndustry.cpp

```diff
@@ -0,0 +1,722 @@
+                    }
+                }
+            }
+
+            // Create new tile
+            if (flags & Flags::apply)
+            {
+                if (!(flags & Flags::ghost))
+                {
+                    World::TileManager::removeSurfaceIndustry(World::toWorldSpace(tilePos));
+                    World::TileManager::setTerrainStyleAsCleared(World::toWorldSpace(tilePos));
+                }
+                auto* elIndustry = World::TileManager::insertElement<World::IndustryElement>(World::toWorldSpace(tilePos), highestBaseZ, 0xF);
+                if (elIndustry == nullptr)
+                {
+                    return FAILURE;
+                }
+                elIndustry->setClearZ(clearZ);
+                elIndustry->setRotation(direction);
+                elIndustry->setIsConstructed(buildImmediate);
+                elIndustry->setIndustryId(industryId);
+                elIndustry->setSequenceIndex(offset.index);
+                // TODO: If it turns out there are more vars in _5 they should be cleared here
+                elIndustry->setSectionProgress(0);
+                elIndustry->setColour(colour);
+                elIndustry->setBuildingType(buildingType);
+                elIndustry->setVar_6_003F(0);
+                World::AnimationManager::createAnimation(3, World::toWorldSpace(tilePos), elIndustry->baseZ());
+                elIndustry->setGhost(flags & Flags::ghost);
+                Ui::ViewportManager::invalidate(World::toWorldSpace(tilePos), elIndustry->baseHeight(), elIndustry->clearHeight());
+            }
+        }
+
+        auto& tileRef = industry->tiles[industry->numTiles];
+        tileRef = World::Pos3(pos, (highestBaseZ * World::kSmallZStep | (isMultiTile ? 0x8000 : 0)));
+        industry->numTiles++;
+        return totalCost;
+    }
+
+    // 0x0045442A
+    static Colour getRandomAvailableColour(Core::Prng& prng, const IndustryObject& indObj)
+    {
+        // Could use a fixed size vector here as its never > Colour::max
+        std::vector<Colour> availableColours;
+        auto colourBitSet = indObj.availableColours;
+        for (auto colourI32 = Numerics::bitScanForward(colourBitSet); colourI32 != -1; colourI32 = Numerics::bitScanForward(colourBitSet))
+        {
+            colourBitSet &= ~(1ULL << colourI32);
+            availableColours.push_back(static_cast<Colour>(colourI32));
+        }
+
+        if (availableColours.empty())
+        {
+            return Colour::black;
+        }
+        // Note: Don't optimise for size 1 as randNext required to prevent divergence
+        return availableColours[prng.randNext(availableColours.size() - 1)];
+    }
+
+    struct RangeAndMinDistance
+    {
+        tile_coord_t distanceRange;
+        tile_coord_t minDistance;
+    };
+
+    // As more placement attempts happen the distance searched to try place increases
+    constexpr std::array<RangeAndMinDistance, 25> kPlacementAttemptDistanceRangeAndMin = {
```

Comment:
Did we introduce a lookup table for this?

Reply:
yes. It was a bit annoying to read the none lookup table version.